File: anti-call-masking/detection-service-rust/src/domain/value_objects.rs
//! Value Objects - Immutable domain primitives with validation
//!
//! These types encapsulate validation rules and are used throughout the domain.

use serde::{Deserialize, Serialize};
use std::fmt;
use std::net::IpAddr;
use std::str::FromStr;
use uuid::Uuid;
use validator::Validate;

use super::errors::DomainError;

/// Unique identifier for a call within the detection system
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct CallId(String);

impl CallId {
    /// Creates a new CallId from a string, validating format
    pub fn new(id: impl Into<String>) -> Result<Self, DomainError> {
        let id = id.into();
        if id.is_empty() {
            return Err(DomainError::InvalidCallId("Call ID cannot be empty".into()));
        }
        if id.len() > 256 {
            return Err(DomainError::InvalidCallId("Call ID too long".into()));
        }
        Ok(Self(id))
    }

    /// Generates a new unique CallId
    pub fn generate() -> Self {
        Self(Uuid::new_v4().to_string())
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl fmt::Display for CallId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

/// Mobile Subscriber Integrated Services Digital Network Number (phone number)
/// Validates Nigerian +234 format and international formats
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct MSISDN(String);

impl MSISDN {
    /// Creates a new MSISDN with validation
    pub fn new(number: impl Into<String>) -> Result<Self, DomainError> {
        let number = Self::normalize(number.into());
        
        if number.is_empty() {
            return Err(DomainError::InvalidMSISDN("MSISDN cannot be empty".into()));
        }
        
        // Must start with + for international format
        if !number.starts_with('+') {
            return Err(DomainError::InvalidMSISDN(
                "MSISDN must be in international format (+...)".into(),
            ));
        }
        
        // Must be digits only after +
        if !number[1..].chars().all(|c| c.is_ascii_digit()) {
            return Err(DomainError::InvalidMSISDN(
                "MSISDN must contain only digits after country code".into(),
            ));
        }
        
        // Minimum length check (country code + subscriber number)
        if number.len() < 8 {
            return Err(DomainError::InvalidMSISDN("MSISDN too short".into()));
        }
        
        if number.len() > 16 {
            return Err(DomainError::InvalidMSISDN("MSISDN too long".into()));
        }
        
        Ok(Self(number))
    }

    /// Normalizes phone number to E.164 format
    fn normalize(mut number: String) -> String {
        // Remove common formatting characters
        number.retain(|c| !matches!(c, ' ' | '-' | '(' | ')' | '.'));
        
        // Convert Nigerian local format to international
        if number.starts_with("234") && !number.starts_with('+') {
            number = format!("+{}", number);
        } else if number.starts_with("0") && number.len() == 11 {
            // Nigerian local format 0XXXXXXXXXX -> +234XXXXXXXXXX
            number = format!("+234{}", &number[1..]);
        }
        
        number
    }

    /// Checks if this is a Nigerian number
    pub fn is_nigerian(&self) -> bool {
        self.0.starts_with("+234")
    }

    /// Returns the country code (without +)
    pub fn country_code(&self) -> &str {
        if self.0.starts_with("+234") {
            "234"
        } else if self.0.len() > 3 {
            &self.0[1..4] // Approximate for other countries
        } else {
            ""
        }
    }

    /// Returns the national number (without country code)
    pub fn national_number(&self) -> &str {
        if self.0.starts_with("+234") {
            &self.0[4..]
        } else {
            &self.0[1..]
        }
    }

    /// Get prefix for MNO identification (first 7 digits for Nigeria)
    pub fn prefix(&self, len: usize) -> &str {
        let national = self.national_number();
        if national.len() >= len {
            &national[..len]
        } else {
            national
        }
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl fmt::Display for MSISDN {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

/// IP Address value object with validation
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct IPAddress(IpAddr);

impl IPAddress {
    /// Creates a new IPAddress from string
    pub fn new(ip: impl AsRef<str>) -> Result<Self, DomainError> {
        let addr = IpAddr::from_str(ip.as_ref())
            .map_err(|_| DomainError::InvalidIPAddress(format!("Invalid IP: {}", ip.as_ref())))?;
        Ok(Self(addr))
    }

    /// Returns true if this is a private/internal IP
    pub fn is_private(&self) -> bool {
        match self.0 {
            IpAddr::V4(ip) => ip.is_private() || ip.is_loopback(),
            IpAddr::V6(ip) => ip.is_loopback(),
        }
    }

    /// Returns true if this is an international gateway IP (non-Nigerian)
    /// This is a heuristic - in production would use GeoIP
    pub fn is_likely_international(&self) -> bool {
        // Placeholder - would integrate with GeoIP service
        !self.is_private()
    }

    pub fn inner(&self) -> &IpAddr {
        &self.0
    }
}

impl fmt::Display for IPAddress {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

/// Fraud confidence score (0.0 - 1.0)
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct FraudScore(f64);

impl FraudScore {
    /// Creates a new FraudScore, clamping to valid range
    pub fn new(score: f64) -> Self {
        Self(score.clamp(0.0, 1.0))
    }

    /// Creates a score from a percentage (0-100)
    pub fn from_percentage(pct: f64) -> Self {
        Self::new(pct / 100.0)
    }

    /// Returns the score as a percentage
    pub fn as_percentage(&self) -> f64 {
        self.0 * 100.0
    }

    /// Determines severity level from score
    pub fn severity(&self) -> Severity {
        match self.0 {
            s if s >= 0.9 => Severity::Critical,
            s if s >= 0.7 => Severity::High,
            s if s >= 0.5 => Severity::Medium,
            s if s >= 0.3 => Severity::Low,
            _ => Severity::Info,
        }
    }

    /// Checks if score exceeds threshold for automatic blocking
    pub fn exceeds_block_threshold(&self) -> bool {
        self.0 >= 0.85
    }

    pub fn value(&self) -> f64 {
        self.0
    }
}

impl Default for FraudScore {
    fn default() -> Self {
        Self(0.0)
    }
}

/// Alert/fraud severity level
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Severity {
    Info = 1,
    Low = 2,
    Medium = 3,
    High = 4,
    Critical = 5,
}

impl Severity {
    pub fn as_int(&self) -> i32 {
        *self as i32
    }
}

/// Fraud detection event types
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum FraudType {
    /// CLI masking - international call with local CLI
    CliMasking,
    /// SIM box - multiple calls from same source
    SimBox,
    /// Call refiling - international to local bypass
    Refiling,
    /// STIR/SHAKEN attestation failure
    StirShakenFail,
    /// Abnormal call pattern
    AnomalousPattern,
    /// Multiple distinct callers to same number
    MaskingAttack,
}

impl fmt::Display for FraudType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::CliMasking => write!(f, "CLI_MASKING"),
            Self::SimBox => write!(f, "SIM_BOX"),
            Self::Refiling => write!(f, "REFILING"),
            Self::StirShakenFail => write!(f, "STIR_SHAKEN_FAIL"),
            Self::AnomalousPattern => write!(f, "ANOMALOUS_PATTERN"),
            Self::MaskingAttack => write!(f, "MASKING_ATTACK"),
        }
    }
}

/// Call status within the detection window
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum CallStatus {
    /// Call is ringing
    Ringing,
    /// Call is active/connected
    Active,
    /// Call completed normally
    Completed,
    /// Call failed/rejected
    Failed,
    /// Call blocked by detection
    Blocked,
}

/// Time window for detection analysis
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct DetectionWindow {
    seconds: u32,
}

impl DetectionWindow {
    pub fn new(seconds: u32) -> Result<Self, DomainError> {
        if seconds == 0 {
            return Err(DomainError::InvalidConfiguration(
                "Detection window must be positive".into(),
            ));
        }
        if seconds > 300 {
            return Err(DomainError::InvalidConfiguration(
                "Detection window cannot exceed 300 seconds".into(),
            ));
        }
        Ok(Self { seconds })
    }

    pub fn default_window() -> Self {
        Self { seconds: 5 }
    }

    pub fn seconds(&self) -> u32 {
        self.seconds
    }
}

impl Default for DetectionWindow {
    fn default() -> Self {
        Self::default_window()
    }
}

/// Threshold for triggering fraud alerts
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct DetectionThreshold {
    distinct_callers: usize,
}

impl DetectionThreshold {
    pub fn new(distinct_callers: usize) -> Result<Self, DomainError> {
        if distinct_callers == 0 {
            return Err(DomainError::InvalidConfiguration(
                "Threshold must be positive".into(),
            ));
        }
        if distinct_callers > 100 {
            return Err(DomainError::InvalidConfiguration(
                "Threshold too high (max 100)".into(),
            ));
        }
        Ok(Self { distinct_callers })
    }

    pub fn default_threshold() -> Self {
        Self { distinct_callers: 5 }
    }

    pub fn distinct_callers(&self) -> usize {
        self.distinct_callers
    }
}

impl Default for DetectionThreshold {
    fn default() -> Self {
        Self::default_threshold()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    mod call_id_tests {
        use super::*;

        #[test]
        fn test_valid_call_id() {
            let id = CallId::new("call-123").unwrap();
            assert_eq!(id.as_str(), "call-123");
        }

        #[test]
        fn test_empty_call_id_fails() {
            assert!(CallId::new("").is_err());
        }

        #[test]
        fn test_generate_unique() {
            let id1 = CallId::generate();
            let id2 = CallId::generate();
            assert_ne!(id1, id2);
        }
    }

    mod msisdn_tests {
        use super::*;

        #[test]
        fn test_valid_nigerian_number() {
            let msisdn = MSISDN::new("+2348012345678").unwrap();
            assert!(msisdn.is_nigerian());
            assert_eq!(msisdn.country_code(), "234");
        }

        #[test]
        fn test_normalizes_local_format() {
            let msisdn = MSISDN::new("08012345678").unwrap();
            assert_eq!(msisdn.as_str(), "+2348012345678");
        }

        #[test]
        fn test_normalizes_without_plus() {
            let msisdn = MSISDN::new("2348012345678").unwrap();
            assert_eq!(msisdn.as_str(), "+2348012345678");
        }

        #[test]
        fn test_invalid_format_fails() {
            assert!(MSISDN::new("invalid").is_err());
            assert!(MSISDN::new("").is_err());
        }

        #[test]
        fn test_prefix_extraction() {
            let msisdn = MSISDN::new("+2348012345678").unwrap();
            assert_eq!(msisdn.prefix(4), "8012");
        }
    }

    mod ip_address_tests {
        use super::*;

        #[test]
        fn test_valid_ipv4() {
            let ip = IPAddress::new("192.168.1.1").unwrap();
            assert!(ip.is_private());
        }

        #[test]
        fn test_valid_ipv6() {
            let ip = IPAddress::new("::1").unwrap();
            assert!(ip.is_private());
        }

        #[test]
        fn test_invalid_ip_fails() {
            assert!(IPAddress::new("not-an-ip").is_err());
        }
    }

    mod fraud_score_tests {
        use super::*;

        #[test]
        fn test_score_clamping() {
            let score = FraudScore::new(1.5);
            assert_eq!(score.value(), 1.0);

            let score = FraudScore::new(-0.5);
            assert_eq!(score.value(), 0.0);
        }

        #[test]
        fn test_severity_levels() {
            assert_eq!(FraudScore::new(0.95).severity(), Severity::Critical);
            assert_eq!(FraudScore::new(0.75).severity(), Severity::High);
            assert_eq!(FraudScore::new(0.55).severity(), Severity::Medium);
            assert_eq!(FraudScore::new(0.35).severity(), Severity::Low);
            assert_eq!(FraudScore::new(0.15).severity(), Severity::Info);
        }

        #[test]
        fn test_block_threshold() {
            assert!(FraudScore::new(0.9).exceeds_block_threshold());
            assert!(!FraudScore::new(0.5).exceeds_block_threshold());
        }
    }

    mod detection_window_tests {
        use super::*;

        #[test]
        fn test_valid_window() {
            let window = DetectionWindow::new(10).unwrap();
            assert_eq!(window.seconds(), 10);
        }

        #[test]
        fn test_zero_window_fails() {
            assert!(DetectionWindow::new(0).is_err());
        }

        #[test]
        fn test_excessive_window_fails() {
            assert!(DetectionWindow::new(500).is_err());
        }
    }
}

File: services/sip-processor/app/domain/value_objects/value_objects.py
"""
Domain Value Objects - Immutable objects with validation

These value objects encapsulate Nigerian telecom-specific validations
and provide type-safe representations used across the domain.
"""

import re
from dataclasses import dataclass
from typing import Optional, Tuple
from enum import Enum
from ipaddress import ip_address, IPv4Address, IPv6Address


class InvalidMSISDNError(ValueError):
    """Raised when MSISDN format is invalid"""
    pass


class InvalidIPAddressError(ValueError):
    """Raised when IP address format is invalid"""
    pass


# Nigerian carrier prefixes mapping
NIGERIAN_PREFIXES = {
    "MTN": ["0803", "0806", "0703", "0706", "0813", "0816", "0814", "0903", "0906"],
    "GLO": ["0805", "0807", "0705", "0815", "0811", "0905"],
    "AIRTEL": ["0802", "0808", "0708", "0812", "0701", "0902", "0901", "0907"],
    "9MOBILE": ["0809", "0818", "0817", "0909", "0908"],
}


@dataclass(frozen=True)
class MSISDN:
    """
    Mobile Station International Subscriber Directory Number
    
    Represents a Nigerian phone number with validation and normalization.
    Format: +234XXXXXXXXXX (E.164)
    """
    value: str
    
    def __post_init__(self):
        if not self._is_valid(self.value):
            raise InvalidMSISDNError(f"Invalid MSISDN format: {self.value}")
    
    @staticmethod
    def _is_valid(value: str) -> bool:
        """Validates Nigerian MSISDN format"""
        # Pattern: +234 followed by 10 digits
        pattern = r'^\+?234[0-9]{10}$'
        return bool(re.match(pattern, value))
    
    @classmethod
    def from_string(cls, value: str) -> 'MSISDN':
        """Creates MSISDN from various formats with normalization"""
        normalized = cls._normalize(value)
        return cls(normalized)
    
    @staticmethod
    def _normalize(value: str) -> str:
        """Normalizes to +234XXXXXXXXXX format"""
        # Remove spaces and dashes
        cleaned = re.sub(r'[\s\-]', '', value)
        
        # Handle different formats
        if cleaned.startswith('0') and len(cleaned) == 11:
            return '+234' + cleaned[1:]
        elif cleaned.startswith('234') and len(cleaned) == 13:
            return '+' + cleaned
        elif cleaned.startswith('+234') and len(cleaned) == 14:
            return cleaned
        
        return cleaned
    
    @property
    def prefix(self) -> str:
        """Returns the 4-digit prefix (e.g., 0803)"""
        if self.value.startswith('+234'):
            return '0' + self.value[4:7]
        return self.value[:4]
    
    @property
    def carrier(self) -> Optional[str]:
        """Returns the carrier name based on prefix"""
        prefix = self.prefix
        for carrier, prefixes in NIGERIAN_PREFIXES.items():
            if prefix in prefixes:
                return carrier
        return None
    
    @property
    def is_nigerian(self) -> bool:
        """Returns True if this is a Nigerian number"""
        return self.value.startswith('+234')
    
    def __str__(self) -> str:
        return self.value


@dataclass(frozen=True)
class IPAddress:
    """
    IP Address value object with validation and classification
    """
    value: str
    _parsed: object = None
    
    def __post_init__(self):
        try:
            parsed = ip_address(self.value)
            object.__setattr__(self, '_parsed', parsed)
        except ValueError:
            raise InvalidIPAddressError(f"Invalid IP address: {self.value}")
    
    @property
    def is_ipv4(self) -> bool:
        return isinstance(self._parsed, IPv4Address)
    
    @property
    def is_ipv6(self) -> bool:
        return isinstance(self._parsed, IPv6Address)
    
    @property
    def is_private(self) -> bool:
        """Returns True if this is a private/internal IP"""
        return self._parsed.is_private
    
    @property
    def is_likely_international(self) -> bool:
        """Heuristic: non-private IPs from non-Nigerian ranges may be international"""
        return not self.is_private
    
    def __str__(self) -> str:
        return self.value


class Severity(Enum):
    """Alert severity levels"""
    LOW = 1
    MEDIUM = 2
    HIGH = 3
    CRITICAL = 4
    
    @classmethod
    def from_score(cls, score: float) -> 'Severity':
        """Determines severity from confidence score"""
        if score >= 0.9:
            return cls.CRITICAL
        elif score >= 0.75:
            return cls.HIGH
        elif score >= 0.5:
            return cls.MEDIUM
        return cls.LOW


class FraudType(Enum):
    """Types of fraud detected"""
    CLI_MASKING = "cli_masking"
    SIMBOX = "simbox"
    WANGIRI = "wangiri"
    IRSF = "irsf"
    PBX_HACKING = "pbx_hacking"
    UNKNOWN = "unknown"


@dataclass(frozen=True)
class FraudScore:
    """
    Fraud confidence score (0.0 - 1.0)
    
    Represents the confidence level of fraud detection.
    """
    value: float
    
    def __post_init__(self):
        # Clamp value between 0 and 1
        clamped = max(0.0, min(1.0, self.value))
        object.__setattr__(self, 'value', clamped)
    
    @property
    def severity(self) -> Severity:
        return Severity.from_score(self.value)
    
    @property
    def is_high_confidence(self) -> bool:
        return self.value >= 0.8
    
    @property
    def exceeds_block_threshold(self) -> bool:
        return self.value >= 0.9
    
    def __float__(self) -> float:
        return self.value


@dataclass(frozen=True)
class CallId:
    """Unique call identifier"""
    value: str
    
    def __post_init__(self):
        if not self.value or len(self.value.strip()) == 0:
            raise ValueError("CallId cannot be empty")
    
    @classmethod
    def generate(cls) -> 'CallId':
        import uuid
        return cls(str(uuid.uuid4()))
    
    def __str__(self) -> str:
        return self.value


@dataclass(frozen=True)
class DetectionWindow:
    """Detection time window configuration"""
    seconds: int
    
    def __post_init__(self):
        if self.seconds < 1 or self.seconds > 60:
            raise ValueError(f"Detection window must be 1-60 seconds, got {self.seconds}")


@dataclass(frozen=True)
class DetectionThreshold:
    """Threshold for distinct caller detection"""
    distinct_callers: int
    
    def __post_init__(self):
        if self.distinct_callers < 2 or self.distinct_callers > 100:
            raise ValueError(f"Threshold must be 2-100 callers, got {self.distinct_callers}")
